lib/hiera/backend.rb

```diff
@@ -196,14 +196,20 @@ def merge_answer(left,right)
       # Backend instances are cached so if you need to connect to any
       # databases then do so in your constructor, future calls to your
       # backend will not create new instances
-      def lookup(key, default, scope, order_override, resolution_type)
+      def lookup(key, default, scope, order_override, resolution_type, recurse_guard = nil)
         @backends ||= {}
         answer = nil
 
         Config[:backends].each do |backend|
           if constants.include?("#{backend.capitalize}_backend") || constants.include?("#{backend.capitalize}_backend".to_sym)
             @backends[backend] ||= Backend.const_get("#{backend.capitalize}_backend").new
-            new_answer = @backends[backend].lookup(key, scope, order_override, resolution_type)
+            backend = @backends[backend]
+            begin
+              new_answer = backend.lookup(key, scope, order_override, resolution_type, recurse_guard)
+            rescue ArgumentError
+              # Handle old style backend that take four arguments
```

Comment:
Hm, this means it will catch argument errors that surface from an existing 5 arg lookup as well. I think it needs to look at the arity of the called lookup method instead. It also needs to issue a debug message when the 4 argument version is called. (Not sure if there is a way to issue a deprecation warning in hiera code base (that does outputs the warning once as opposed to on each call to lookup).